shared/js/background/dnr-session-rule-id.js

```diff
@@ -0,0 +1,20 @@
+/**
+ * For managing dynamically created MV3 session rules
+ * getNextSessionRuleId will return the next unique session rule id to use when creating new session rules
+ **/
+const browserWrapper = require('./wrapper.es6')
+const SESSION_RULE_ID_START = 100000
+const SESSION_RULE_STORAGE_KEY = 'sessionRuleOffset'
+let sessionRuleOffset = 0
+browserWrapper.getFromSessionStorage(SESSION_RULE_STORAGE_KEY).then(offset => {
+    if (offset) {
+        sessionRuleOffset = offset
+    }
+})
+
+export function getNextSessionRuleId () {
+    const nextRuleId = SESSION_RULE_ID_START + sessionRuleOffset
+    sessionRuleOffset += 1
```

Comment:
This should await the getting from storage.

Reply:
Can we stop the tab from loading while we wait?

Reply:
No so you probably want to precalculate this and continue to have this method synchronous.

Reply:
We're reading the ID offset from session storage when the background ServiceWorker starts, so that `getNextSessionRuleId` can return synchronously. We're doing that to avoid the race condition where the allowing DNR rule is not added quickly enough to ensure the corresponding request isn't blocked.

Reply:
I think we're saying the same thing here.

We don't want an await on what we call sync in a tab I agree; however we should await this somewhere otherwise we risk overriding the id. We need an async queue to block getting the next value (somewhere) otherwise two writes could write different values to the store.

The two issues as I see them:

1. `getNextSessionRuleId` could be triggered before the `sessionRuleOffset` is called.
2. `getNextSessionRuleId` could be called twice whilst `setToSessionStorage` happens resulting in an off by 1 error.

Perhaps the risk of 2. happening and a service worker shut down is too low to be bothered with the defensive code? 1. however seems pretty likely.